cnd/src/btsieve/bitcoin/bitcoin_cache.rs

```diff
@@ -0,0 +1,94 @@
+use crate::btsieve::{BlockByHash, LatestBlock};
+use bitcoin::{hashes::sha256d, util::hash::BitcoinHash};
+use fmt::Debug;
+use futures::Future;
+use futures_core::{
+    compat::Future01CompatExt,
+    future::{FutureExt, TryFutureExt},
+};
+use std::{fmt, fmt::Formatter, sync::Arc};
+use tokio::sync::Mutex;
+
+pub struct BitcoinCache<T> {
+    pub inner: T,
+    pub cache: Arc<Mutex<lru::LruCache<sha256d::Hash, bitcoin::Block>>>,
+}
+
+impl<T> Clone for BitcoinCache<T>
+where
+    T: Clone,
+{
+    fn clone(&self) -> Self {
+        BitcoinCache {
+            inner: self.inner.clone(),
+            cache: self.cache.clone(),
+        }
+    }
+}
+
+impl<T> Debug for BitcoinCache<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        self.inner.fmt(f)
+    }
+}
+
+impl<T> LatestBlock for BitcoinCache<T>
+where
+    T: LatestBlock<Block = bitcoin::Block, BlockHash = sha256d::Hash> + Clone,
+{
+    type Block = bitcoin::Block;
+    type BlockHash = sha256d::Hash;
+
+    fn latest_block(
+        &mut self,
+    ) -> Box<dyn Future<Item = Self::Block, Error = anyhow::Error> + Send + 'static> {
+        let cache = self.cache.clone();
+        let mut inner = self.inner.clone();
+
+        let future = async move {
+            let block = inner.latest_block().compat().await?;
+            let mut guard = cache.lock().await;
+            guard.put(block.bitcoin_hash(), block.clone());
+            Ok(block)
+        }
+        .boxed()
+        .compat();
+
+        Box::new(future)
+    }
+}
+
+impl<T> BlockByHash for BitcoinCache<T>
+where
+    T: BlockByHash<Block = bitcoin::Block, BlockHash = sha256d::Hash> + Clone,
+{
+    type Block = bitcoin::Block;
+    type BlockHash = sha256d::Hash;
+
+    fn block_by_hash(
+        &self,
+        block_hash: Self::BlockHash,
+    ) -> Box<dyn Future<Item = Self::Block, Error = anyhow::Error> + Send + 'static> {
+        let cache = self.cache.clone();
+        let inner = self.inner.clone();
+
+        let future = async move {
+            match cache.lock().await.get(&block_hash) {
+                Some(block) => Ok(block.clone()),
+                None => {
+                    let block = inner.block_by_hash(block_hash.clone()).compat().await?;
+                    let mut guard = cache.lock().await;
+                    guard.put(block_hash, block.clone());
```

Comment:
We take the lock twice here, perhaps it could be written like:
```
diff --git a/cnd/src/btsieve/bitcoin/bitcoin_cache.rs b/cnd/src/btsieve/bitcoin/bitcoin_cache.rs
index 3f82bda4..ea9d6f7e 100644
--- a/cnd/src/btsieve/bitcoin/bitcoin_cache.rs
+++ b/cnd/src/btsieve/bitcoin/bitcoin_cache.rs
@@ -70,12 +70,12 @@ where
         let inner = self.inner.clone();
 
         let future = async move {
-            match cache.lock().await.get(&block_hash) {
+            let mut cache = cache.lock().await;
+            match cache.get(&block_hash) {
                 Some(block) => Ok(block.clone()),
                 None => {
                     let block = inner.block_by_hash(block_hash.clone()).compat().await?;
-                    let mut guard = cache.lock().await;
-                    guard.put(block_hash, block.clone());
+                    cache.put(block_hash, block.clone());
                     Ok(block)
                 }
             }
```

Reply:
In non-diff form
```
        let future = async move {
            let mut cache = cache.lock().await;
            match cache.get(&block_hash) {
                Some(block) => Ok(block.clone()),
                None => {
                    let block = inner.block_by_hash(block_hash.clone()).compat().await?;
                    cache.put(block_hash, block.clone());
                    Ok(block)
                }
            }
        }
```

Reply:
If we only take the lock once, we have it open for too long, including one async operation which might timeout, blocking all other requests.

It was on purpose to only lock for as short as possible :)